### PR TITLE
enable flat directory certificate store for k8s cert-manager compatibility

### PR DESCRIPTION
**Pull request by cristipogacean**
## Purpose
* add support for flat directory certificate store for k8s cert manager compatibility
* handle .crt and .key certificate types

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code

```
```

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

File: src/OpcApplicationConfiguration.cs [hunk 1]
```diff
@@ -56,12 +57,31 @@ public static string Hostname
...
+    /// </summary>
+    private static bool _flatDirectoryCertStoreInitialized = false;
 
     /// <summary>
     /// Configures all OPC stack settings.
     /// </summary>
     public async Task<ApplicationConfiguration> ConfigureAsync()
-    {
+    {
+        if (!_flatDirectoryCertStoreInitialized)
```
Comment (mregen) [inline src/OpcApplicationConfiguration.cs:76]: isn't this ever called once?

**Review by mregen — APPROVED** [review]

File: src/OpcApplicationConfiguration.cs (hunk 1 above)
Comment (cristipogacean) [reply src/OpcApplicationConfiguration.cs:76]: It should normally be only called once. In the unit tests it is called for every test though.

File: README.md [hunk 2]
```diff
@@ -277,6 +277,7 @@ X.509 certificates:
 * Running on Windows natively, you cannot use an application certificate store of type `Directory`, since the access to the private key will fail. Use the option `--at X509Store` in this case.
 * Running as Linux Docker container, you can map the certificate stores to the host file system by using the Docker run option `-v <hostpkidirectory>:/app/pki`. This will make the certificate persistent over starts.
 * Running as Linux Docker container using an X509Store for the application certificate, you need to use the Docker run option `-v x509certstores:/root/.dotnet/corefx/cryptography/x509stores` and the application option `--at X509Store`
+* When running in kubernetes context, use option `--at FlatDirectory`. This enables the OPC UA server to consume both public key and private key certificates directly from the /app/pki/own/ path without expecting the `certs` and `private` subdirectories. furthermore, certificates with .crt and .key are accepted.
```
Comment (luiscantero) [inline README.md:280]: >furthermore [](http://example.com/codeflow?start=255&length=11)

Use capital

Comment (luiscantero) [inline README.md:280]: >with [](http://example.com/codeflow?start=281&length=4)

Maybe better "of type", or "with the extension"

File: src/FlatDirectoryCertificateStore.cs [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+                {
+                    using X509Certificate2 certificate = new X509Certificate2(file.FullName);
+                    if (!MatchCertificate(certificate, thumbprint, subjectName))
+                    {
+                        continue;
+                    }
+
+                    X509Certificate2 privateKeyCertificate = X509Certificate2.CreateFromPemFile(file.FullName, keyFile.FullName);
+
+                    Utils.LogInfo("Loading private key succeeded for {0} - {1}", thumbprint, subjectName);
```
Comment (luiscantero) [inline src/FlatDirectoryCertificateStore.cs:195]: >{0} - {1} [](http://example.com/codeflow?start=69&length=9)

If possible, could you pls change the numeric placeholders in this file to use strings instead? they get flagged by the analyzers and we need to migrate the remaining ones someday anyway.

File: src/OpcApplicationConfiguration.cs [hunk 4]
```diff
@@ -56,12 +57,31 @@ public static string Hostname
...
     /// </summary>
     public static int OpcMaxStringLength { get; set; } = 4 * 1024 * 1024;
+
+    /// <summary>
+    /// Set when a flat directory certificate store should be used.
+    /// </summary>
+    public static bool EnableFlatDirectoryCertStore { get; set; } = false;
+
+    /// <summary>
+    /// the flat directory certificate store shall only be initialized once
```
Comment (luiscantero) [inline src/OpcApplicationConfiguration.cs:67]: >the [](http://example.com/codeflow?start=8&length=3)

Nit: Capital and end sentence with dot

File: src/OpcApplicationConfiguration.cs [hunk 5]
```diff
@@ -56,12 +57,31 @@ public static string Hostname
...
 
     /// <summary>
     /// Configures all OPC stack settings.
     /// </summary>
     public async Task<ApplicationConfiguration> ConfigureAsync()
-    {
+    {
+        if (!_flatDirectoryCertStoreInitialized)
+        {
+            // Register FlatDirectoryCertificateStoreType as knows certificate store type.
```
Comment (luiscantero) [inline src/OpcApplicationConfiguration.cs:78]: >knows [](http://example.com/codeflow?start=61&length=5)

Typo

File: src/OpcApplicationConfiguration.cs (hunk 1 above)
Comment (luiscantero) [reply src/OpcApplicationConfiguration.cs:76]: btw: Would it be difficult to add unit tests for this? can be done in separate PR.

File: src/FlatDirectoryCertificateStore.cs [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        if (!_innerStore.Directory.Exists)
+        {
+            return await _innerStore.LoadPrivateKey(thumbprint, subjectName, password).ConfigureAwait(false);
+        }
+
+        foreach (FileInfo file in _innerStore.Directory.GetFiles('*' + CrtExtension))
+        {
+            try
+            {
+                FileInfo keyFile = new FileInfo(file.FullName.Replace(CrtExtension, KeyExtension, StringComparison.OrdinalIgnoreCase));
```
Comment (luiscantero) [inline src/FlatDirectoryCertificateStore.cs:184]: >FileInfo [](http://example.com/codeflow?start=16&length=8)

var

File: src/FlatDirectoryCertificateStore.cs [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+        if (!_innerStore.Directory.Exists)
+        {
+            return certificatesCollection;
+        }
+
+        foreach (FileInfo file in _innerStore.Directory.GetFiles('*' + CrtExtension))
+        {
+            try
+            {
+                X509Certificate2Collection certificates = new X509Certificate2Collection();
```
Comment (luiscantero) [inline src/FlatDirectoryCertificateStore.cs:91]: >X509Certificate2Collection [](http://example.com/codeflow?start=16&length=26)

var

File: src/FlatDirectoryCertificateStore.cs [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+        }
+
+        foreach (FileInfo file in _innerStore.Directory.GetFiles('*' + CrtExtension))
+        {
+            try
+            {
+                FileInfo keyFile = new FileInfo(file.FullName.Replace(CrtExtension, KeyExtension, StringComparison.OrdinalIgnoreCase));
+                if (keyFile.Exists)
+                {
+                    using X509Certificate2 certificate = new X509Certificate2(file.FullName);
```
Comment (luiscantero) [inline src/FlatDirectoryCertificateStore.cs:187]: >X509Certificate2 [](http://example.com/codeflow?start=26&length=16)

var

Comment (cristipogacean) [reply src/FlatDirectoryCertificateStore.cs:187]: done

File: src/FlatDirectoryCertificateStore.cs (hunk 3 above)
Comment (cristipogacean) [reply src/FlatDirectoryCertificateStore.cs:195]: done

File: src/FlatDirectoryCertificateStore.cs [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+        if (!_innerStore.Directory.Exists)
+        {
+            return certificatesCollection;
+        }
+
+        foreach (FileInfo file in _innerStore.Directory.GetFiles('*' + CrtExtension))
+        {
+            try
+            {
+                X509Certificate2Collection certificates = new X509Certificate2Collection();
```
[the remark this answers is not in the document]
Comment (cristipogacean) [reply src/FlatDirectoryCertificateStore.cs:146]: done

File: src/FlatDirectoryCertificateStore.cs (hunk 7 above)
Comment (cristipogacean) [reply src/FlatDirectoryCertificateStore.cs:91]: done

File: src/OpcApplicationConfiguration.cs (hunk 4 above)
Comment (cristipogacean) [reply src/OpcApplicationConfiguration.cs:67]: done

File: src/OpcApplicationConfiguration.cs (hunk 5 above)
Comment (cristipogacean) [reply src/OpcApplicationConfiguration.cs:78]: done

File: src/FlatDirectoryCertificateStore.cs (hunk 6 above)
Comment (cristipogacean) [reply src/FlatDirectoryCertificateStore.cs:184]: done

File: src/FlatDirectoryCertificateStore.cs (hunk 3 above)
Comment (cristipogacean) [reply src/FlatDirectoryCertificateStore.cs:195]: done

**Review by luiscantero — APPROVED** [review]
:shipit:

**Pull request merged**